ui/kotlinx-coroutines-android/src/HandlerDispatcher.kt

```diff
@@ -190,11 +190,14 @@ public suspend fun awaitFrame(): Long {
             postFrameCallback(choreographer, cont)
         }
     }
-    // post into looper thread to figure it out
     return suspendCancellableCoroutine { cont ->
-        Dispatchers.Main.dispatch(EmptyCoroutineContext, Runnable {
+        if (Looper.myLooper() === Looper.getMainLooper()) { // Check if we are already in the main looper thread
```

Comment:
Thanks! 

I'm not yet convinced this should be **the** solution though.
I would expect that `Dispatchers.Main.immediate` would be enough, but it is not due to our implementation details.

We'll fix it in the next release, but it is an open question whether we are going to use this particular approach though

Reply:
No probs!

The only reason I took the `Looper` approach here is because this code is specific to Android, and I didn't see anything similar for JavaFX or Swing implementations (`awaitPulse()` was the closest parallel). I see no problems and I am not opposed to using `Dispatchers.Main.immediate`.

Reply:
It seems to be the way though!

I've opened https://github.com/Kotlin/kotlinx.coroutines/pull/3437 with a few changes on top of yours that I'll squash into your commit. Thanks for the contribution, we'll merge it in the next release!